# Some required R packages
library(numbat)
library(dplyr)
library(Seurat)
library(ggplot2)
library(glue)
library(data.table)
library(ggtree)
library(stringr)
library(tidygraph)
library(patchwork)
get_field <- function(string,field=1,delim="_", fixed=T) return(strsplit(string,delim, fixed=fixed)[[1]][field])

# Parallel Computing
library(future)
plan("multicore", workers = 8)
options(future.globals.maxSize= 50 * 1024^3)
options(future.rng.onMisuse = "ignore")

pat <- commandArgs(TRUE)[1]

# Step1 : Reload and Pre-filter cells
print( paste0("Start run of Step1 : ",pat) )

###
seu <- readRDS( paste0("BrM.", pat, ".scObject.rds") )
seu$barcode_new <- rownames(seu@meta.data)
seu$barcode_orig <- paste0(sapply(colnames(seu), get_field, 4, "_"),"-1")

count_mat <- seu@assays$RNA@counts
colnames(count_mat) <- seu$barcode_orig

# allele dataframe generated by pileup_and_phase script
df_allele <- read.delim(paste0('/Numbat/pileup_and_phase/',pat,'/',pat,'_allele_counts.tsv.gz'))
df_allele <- subset(df_allele, cell %in% seu$barcode_orig)

# ref_internal
ref_internal <- readRDS("/Numbat/build_lambdas_ref/numbat.ref_internal.rds")

# Step2 : run Numbat
print( paste0("Start run of Step2 : ",pat) )

out = run_numbat(count_mat=count_mat, # gene x cell integer UMI count matrix 
                 lambdas_ref=ref_internal, # reference expression profile, a gene x cell type normalized expression level matrix
                 df_allele=df_allele, # allele dataframe generated by pileup_and_phase script
                 out_dir= pat,
                 genome = "hg38",
                 ncores = 16,
                 plot = TRUE)

# Step3 : plot Numbat
print( paste0("Start run of Step3 : ",pat) )

mypal = c('1' = 'gray', '2' = "#377EB8", '3' = "#4DAF4A", '4' = "#984EA3", 
          '5'="#ff9768", '6'='#ae1717', '7'='#0f04b5', '8'="#f87382",
          '9'='green','10'='yellow','11'='deeppink','12'='gold4')

# Read-in numbat output
nb = Numbat$new(out_dir = paste0('./',pat))

# Single-cell CNV calls
cnv_calls <- nb$joint_post %>% select(cell, CHROM, seg, cnv_state, p_cnv, p_cnv_x, p_cnv_y)
table(cnv_calls$cnv_state)
cnv_calls %>% group_by(cnv_state) %>% arrange(p_cnv,desc=F)

# Clone info
clones <- dim(table(nb$clone_post$clone_opt))
clone_info <- nb$clone_post
seu$cell <- seu$barcode_orig
seu@meta.data <- left_join(seu@meta.data, clone_info, by='cell')
rownames(seu@meta.data) <- seu$barcode_new

### Plots
pdf(paste0("./",pat,"/plots_",pat,"_numbat.pdf"),width = 10)

# (1) Copy number landscape and single-cell phylogeny
nb$plot_phylo_heatmap(clone_bar = TRUE, p_min = 0.9,raster = T,pal_clone = mypal)

# (2) Consensus copy number segments
nb$plot_consensus()

# (3) Bulk CNV profiles
nb$bulk_clones %>% 
  filter(n_cells > 50) %>%
  plot_bulks(min_LLR = 10, # filtering CNVs by evidence
             legend = TRUE,raster=T)

# (4) clones
g1 <- DimPlot(seu, group.by = 'FinalCellType', shuffle = T, raster=T, label=T, label.size=4)
g2 <- DimPlot(seu, group.by = 'clone_opt', shuffle = T, raster=T, cols = mypal[1:clones])
g3 <- DimPlot(seu, group.by = 'GT_opt',shuffle = T, raster=T)
print((g1+g2+g3)+plot_layout(ncol=2))

# (5) Tumor versus normal probability
p1 <- FeaturePlot(seu, features  = c('p_cnv'),order = T, raster=T)+
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')+
  ggtitle('Tumor vs normal probability\n(joint)')

p2 <- FeaturePlot(seu, features  = c('p_cnv_x'),order = T, raster=T)+
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')+
  ggtitle('Tumor vs normal probability\n(gex)')

p3 <- FeaturePlot(seu, features  = c('p_cnv_y'),order = T, raster=T)+
  scale_color_gradient2(low = 'royalblue', mid = 'white', high = 'red3', midpoint = 0.5, name = 'Posterior')+
  ggtitle('Tumor vs normal probability\n(allele)')
print((p1+p2+p3)+plot_layout(ncol=2))

# (6) Tumor phylogeny
nb$plot_sc_tree(
  label_size = 3, 
  branch_width = 0.5, 
  tip_length = 0.5,
  pal_clone = mypal,
  tip = TRUE)

# (7) mutational history
nb$plot_mut_history(pal=mypal)

dev.off()
